src/desktop/space/element.rs

```diff
@@ -10,6 +10,82 @@ use std::{
 };
 use wayland_server::protocol::wl_surface::WlSurface;
 
+/// Enum for indicating on with layer a render element schould be draw
+#[derive(Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum RenderLayer {
+    /// Bellow every other elements
+    Bottom,
+    /// Above WlrLayer::Background but bellow WlrLayer::Bottom
+    AboveBackground,
+    /// Right before programs windows are draw
+    BeforeWindows,
+    /// Right after programs windows are draw
+    AfterWindows,
+    /// Above WlrLayer::Top but bellow WlrLayer::Overlay
+    BeforeOverlay,
+    /// Above anything else
+    Top,
+}
+
```

Comment:
Between that and the layer shell layers, this is getting pretty heavy. At this point I'm wondering if using a numeric z-index might make more sense?

Basically, assign a numeric value to each defined protocol layer (so maybe `background = 10, bottom = 20, shell = 30, top = 40, overlay = 50` for example), and when drawing elements, we just sort them according to that z-index using a *stable* sort (to preserve the ordering within a layer) and draw that?

This would give a lot of flexibility to the downstream users of smithay when drawing their content, hopefully with a reasonably small runtime cost (collect the render elements in a Vec and sort it). What do you think @Drakulix?

Reply:
Oh right. I mean, that does not cover any use case you cannot already do with the enum.

But I guess downstream would not have to carefully manage the order of each and every custom element, if they could just use a z-index.

Reply:
That is an interesting and flexible idea, but would definitely need more changes in the whole render_output

Reply:
Not necessarily. You could just sort the array, when we create the DynamicList (and we would need to pass a Vec instead of a slice to do that without copying the references) and then filter for indices between a given range instead of layers.